gen/ir/external.go

```diff
@@ -0,0 +1,131 @@
+package ir
+
+import (
+	"go/types"
+	"strings"
+
+	"github.com/go-faster/errors"
+	"golang.org/x/tools/go/packages"
+)
+
+// ExternalType represents an external type.
+type ExternalType struct {
+	PackagePath string
+	GoName      string
+	ImportAlias string
+	Encode      ExternalEncoding
+	Decode      ExternalEncoding
+}
+
+// ExternalEncoding is a kind of external type.
+type ExternalEncoding int
+
+// String returns the string representation of the ExternalEncoding.
+func (e ExternalEncoding) String() string {
+	var s strings.Builder
+	for i := ExternalEncoding(1); i <= e; i <<= 1 {
+		if e.Has(i) {
+			if s.Len() > 0 {
+				s.WriteString("|")
+			}
+			switch i {
+			case ExternalJSON:
+				s.WriteString("JSON")
+			case ExternalText:
+				s.WriteString("Text")
+			case ExternalBinary:
+				s.WriteString("Binary")
+			}
+		}
+	}
+	return s.String()
+}
+
+// Has checks if the encoding kind is present in the ExternalEncoding.
+func (e ExternalEncoding) Has(kind ExternalEncoding) bool {
+	return e&kind != 0
+}
+
+const (
+	// ExternalNative indicates that the type implements ogen's json.Marshaler or json.Unmarshaler.
+	ExternalNative ExternalEncoding = 1 << iota
+	// ExternalJSON indicates that the type implements stdlib json.Marshaler or json.Unmarshaler.
+	ExternalJSON
+	// ExternalText indicates that the type implements stdlib encoding.TextMarshaler or encoding.TextUnmarshaler.
+	ExternalText
+	// ExternalBinary indicates that the type implements stdlib encoding.BinaryMarshaler or encoding.BinaryUnmarshaler.
+	ExternalBinary
```

Comment:
This currently isn't used. I was however considering adding support for `encoding.BinaryMarshaler` and `encoding. BinaryUnmarshaler` for values that have `type: string` and `format: byte`. What do you think?